library/src/main/java/com/getkeepsafe/relinker/ApkLibraryInstaller.java

```diff
@@ -0,0 +1,162 @@
+                ZipEntry libraryEntry = null;
+
+                for (final String abi : abis) {
+                    jniNameInApk = "lib" + File.separatorChar + abi + File.separatorChar
+                            + mappedLibraryName;
+                    libraryEntry = zipFile.getEntry(jniNameInApk);
+
+                    if (libraryEntry != null) {
+                        break;
+                    }
+                }
+
+                if (jniNameInApk != null) instance.log("Looking for %s in APK...", jniNameInApk);
+
+                if (libraryEntry == null) {
+                    // Does not exist in the APK
+                    if (jniNameInApk != null) {
+                        throw new MissingLibraryException(jniNameInApk);
+                    } else {
+                        throw new MissingLibraryException(mappedLibraryName);
+                    }
+                }
+
+                instance.log("Found %s! Extracting...", jniNameInApk);
+                try {
+                    if (!destination.exists() && !destination.createNewFile()) {
+                        continue;
+                    }
+                } catch (IOException ignored) {
+                    // Try again
+                    continue;
+                }
+
+                InputStream inputStream = null;
+                FileOutputStream fileOut = null;
+                try {
+                    inputStream = zipFile.getInputStream(libraryEntry);
+                    fileOut = new FileOutputStream(destination);
+                    copy(inputStream, fileOut);
+                } catch (FileNotFoundException e) {
+                    // Try again
+                    continue;
+                } catch (IOException e) {
+                    // Try again
+                    continue;
+                } finally {
+                    closeSilently(inputStream);
+                    closeSilently(fileOut);
+                }
+
+                // Change permission to rwxr-xr-x
+                destination.setReadable(true, false);
+                destination.setExecutable(true, false);
+                destination.setWritable(true);
+                break;
+            }
+        } finally {
+            try {
+                if (zipFile != null) {
+                    zipFile.close();
+                }
+            } catch (IOException ignored) {}
+        }
+    }
+
+    /**
+     * Copies all data from an {@link InputStream} to an {@link OutputStream}.
+     *
+     * @param in The stream to read from.
+     * @param out The stream to write to.
+     * @throws IOException when a stream operation fails.
+     */
+    private void copy(InputStream in, OutputStream out) throws IOException {
+        byte[] buf = new byte[COPY_BUFFER_SIZE];
+        while (true) {
+            int read = in.read(buf);
+            if (read == -1) {
+                break;
+            }
+            out.write(buf, 0, read);
```

Comment:
use the technique we use to double check that the written output file has the same amount of bytes after you're done. If not, show IOException manually/